File: matlab/src/+kinetic_mhd_pinn/+common/registry.m
function specs = registry()
%REGISTRY Return the project module registry.

specs = repmat(struct('id', '', 'number', 0, 'name', '', 'package', '', ...
    'dependencies', {{}}, 'status', ''), 1, 11);
    specs(1) = makeSpec('module_1', 1, 'Stellarator Equilibrium and Magnetic Geometry', 'module1_stellarator_equilibrium_geometry', {}, 'scaffold');
    specs(2) = makeSpec('module_2', 2, 'Energetic-Particle Sources and Distribution Functions', 'module2_energetic_particle_sources', {'module_1'}, 'scaffold');
    specs(3) = makeSpec('module_3', 3, 'Guiding-Center Orbits, Confinement, and Wall Losses', 'module3_guiding_center_orbits_wall_losses', {'module_1', 'module_2'}, 'scaffold');
    specs(4) = makeSpec('module_4', 4, 'Alfven Continuum and Fluid-MHD Eigenmodes', 'module4_alfven_continuum_eigenmodes', {'module_1', 'module_7'}, 'reduced-benchmark');
    specs(5) = makeSpec('module_5', 5, 'Kinetic Energetic-Particle Drive and Damping', 'module5_kinetic_energetic_particle_drive', {'module_2', 'module_3', 'module_4'}, 'scaffold');
    specs(6) = makeSpec('module_6', 6, 'Alfven-Eigenmode-Induced Fast-Ion Transport', 'module6_ae_induced_fast_ion_transport', {'module_2', 'module_4', 'module_5'}, 'scaffold');
    specs(7) = makeSpec('module_7', 7, 'Alpha Heating and Thermal-Profile Evolution', 'module7_alpha_heating_profile_evolution', {'module_2', 'module_3', 'module_6'}, 'scaffold');
    specs(8) = makeSpec('module_8', 8, 'Coupled Kinetic-MHD Multiphysics Integration', 'module8_coupled_multiphysics_integration', {'module_1', 'module_2', 'module_3', 'module_4', 'module_5', 'module_6', 'module_7'}, 'scaffold');
    specs(9) = makeSpec('module_9', 9, 'PINN and Neural-Operator Surrogates', 'module9_pinn_neural_operator_surrogates', {'module_8'}, 'scaffold');
    specs(10) = makeSpec('module_10', 10, 'Synthetic Diagnostics and Physics-Informed Inverse Modeling', 'module10_synthetic_diagnostics_inverse_modeling', {'module_3', 'module_4', 'module_5', 'module_6', 'module_7', 'module_8'}, 'scaffold');
    specs(11) = makeSpec('module_11', 11, 'GNN Physical Dependency Discovery and Coupling Optimization', 'module11_gnn_dependency_coupling_optimization', {'module_8'}, 'scaffold');
end

function spec = makeSpec(id, number, name, package, dependencies, status)
spec = struct('id', id, 'number', number, 'name', name, ...
    'package', package, 'dependencies', {dependencies}, 'status', status);
end

File: python/src/kinetic_mhd_pinn/common/registry.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ModuleSpec:
    id: str
    number: int
    name: str
    package: str
    dependencies: tuple[str, ...]
    status: str


MODULE_SPECS: tuple[ModuleSpec, ...] = (
    ModuleSpec(id='module_1', number=1, name='Stellarator Equilibrium and Magnetic Geometry', package='module1_stellarator_equilibrium_geometry', dependencies=(), status='scaffold'),
    ModuleSpec(id='module_2', number=2, name='Energetic-Particle Sources and Distribution Functions', package='module2_energetic_particle_sources', dependencies=('module_1',), status='scaffold'),
    ModuleSpec(id='module_3', number=3, name='Guiding-Center Orbits, Confinement, and Wall Losses', package='module3_guiding_center_orbits_wall_losses', dependencies=('module_1', 'module_2'), status='scaffold'),
    ModuleSpec(id='module_4', number=4, name='Alfvén Continuum and Fluid-MHD Eigenmodes', package='module4_alfven_continuum_eigenmodes', dependencies=('module_1', 'module_7'), status='reduced-benchmark'),
    ModuleSpec(id='module_5', number=5, name='Kinetic Energetic-Particle Drive and Damping', package='module5_kinetic_energetic_particle_drive', dependencies=('module_2', 'module_3', 'module_4'), status='scaffold'),
    ModuleSpec(id='module_6', number=6, name='Alfvén-Eigenmode-Induced Fast-Ion Transport', package='module6_ae_induced_fast_ion_transport', dependencies=('module_2', 'module_4', 'module_5'), status='scaffold'),
    ModuleSpec(id='module_7', number=7, name='Alpha Heating and Thermal-Profile Evolution', package='module7_alpha_heating_profile_evolution', dependencies=('module_2', 'module_3', 'module_6'), status='scaffold'),
    ModuleSpec(id='module_8', number=8, name='Coupled Kinetic-MHD Multiphysics Integration', package='module8_coupled_multiphysics_integration', dependencies=('module_1', 'module_2', 'module_3', 'module_4', 'module_5', 'module_6', 'module_7'), status='scaffold'),
    ModuleSpec(id='module_9', number=9, name='PINN and Neural-Operator Surrogates', package='module9_pinn_neural_operator_surrogates', dependencies=('module_8',), status='scaffold'),
    ModuleSpec(id='module_10', number=10, name='Synthetic Diagnostics and Physics-Informed Inverse Modeling', package='module10_synthetic_diagnostics_inverse_modeling', dependencies=('module_3', 'module_4', 'module_5', 'module_6', 'module_7', 'module_8'), status='scaffold'),
    ModuleSpec(id='module_11', number=11, name='GNN Physical Dependency Discovery and Coupling Optimization', package='module11_gnn_dependency_coupling_optimization', dependencies=('module_8',), status='scaffold'),
)


def nominal_startup_order() -> tuple[str, ...]:
    """Return the architecture's nominal initialization order.

    The eventual coupled time integrator may iterate feedback groups in a different
    order; this function only defines deterministic startup and smoke-test order.
    """
    return tuple(spec.id for spec in sorted(MODULE_SPECS, key=lambda item: item.number))


def validate_registry() -> None:
    ids = [spec.id for spec in MODULE_SPECS]
    if len(ids) != len(set(ids)):
        raise ValueError("Module IDs must be unique")
    known = set(ids)
    for spec in MODULE_SPECS:
        unknown = set(spec.dependencies) - known
        if unknown:
            raise ValueError(f"{spec.id} has unknown dependencies: {sorted(unknown)}")
